// @clerk/shared/dist/saml.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/saml.ts
var saml_exports = {};
__export(saml_exports, {
  SAML_IDPS: () => SAML_IDPS
});
module.exports = __toCommonJS(saml_exports);
var SAML_IDPS = {
  saml_okta: {
    name: "Okta Workforce",
    logo: "okta"
  },
  saml_google: {
    name: "Google Workspace",
    logo: "google"
  },
  saml_microsoft: {
    name: "Microsoft Entra ID (Formerly AD)",
    logo: "azure"
  },
  saml_custom: {
    name: "SAML",
    logo: "saml"
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  SAML_IDPS
});
